### PR TITLE
Add support for building with ghc-9.2 and CHaP readiness

### PR DESCRIPTION
**Pull request by erikd**

### CODE REVIEW

**Review by deepfire — APPROVED** [review]
Thank you a lot, @erikd !

File: cabal.project [hunk 1]
```diff
@@ -34,9 +30,16 @@ source-repository-package
   --sha256: 0pb7bg0936fldaa5r08nqbxvi2g8pcy4w3c7kdcg7pdgmimr30ss
 
 constraints:
-  ip < 1.5,
   hedgehog >= 1.0,
-  bimap >= 0.4.0
+  bimap >= 0.4.0,
+  bytestring >= 0.11
```
Comment (michaelpj) [inline cabal.project:35]: The bytestring constraint should go in the cabal file of the package that needs it, perhaps? The others could do with comments so we know why we need them and when we can remove them?

File: cabal.project [hunk 2]
```diff
@@ -34,9 +30,16 @@ source-repository-package
...
+
+allow-newer:
+    libsystemd-journal:base
+  , ekg:*
+  , ekg-core:base
+  , ekg-json:base
+  , websockets-snap:bytestring
```
Comment (michaelpj) [inline cabal.project:42]: ditto on the comments

**Review by michaelpj — COMMENTED** [review]
CI is stuck on Windows because the version of bytestring needed by Win32-network conflicts. Looks like you can fix that by bumping the Win32-network dep, newer ones have `bytestring < 0.12` instead.

File: .github/workflows/haskell.yml [hunk 3]
```diff
@@ -1,73 +1,178 @@
...
+      run: echo "cabal-store=/home/runner/.cabal/store" >> $GITHUB_OUTPUT
+
+    - name: "OUTPUT Record cabal-store (Windows)"
+      id: win-setup-haskell
+      if: runner.os == 'Windows'
+      shell: bash
+      run: echo "cabal-store=C:\\cabal\\store" >> $GITHUB_OUTPUT
+
+    - name: Set cache version
+      run: echo "CACHE_VERSION=grFfw7r" >> $GITHUB_ENV
```
Comment (angerman) [inline .github/workflows/haskell.yml:93]: This one probably needs a different name, not to clash with other caches.

File: .github/workflows/haskell.yml [hunk 4]
```diff
@@ -1,73 +1,178 @@
...
 
-    - name: Cabal Configure
-      run: retry 3 cabal configure --enable-tests --enable-benchmarks --write-ghc-environment-files=always
+    - name: Configure build
+      shell: bash
+      run: |
+        if [ "${{github.event.inputs.tests}}" == "all" ]; then
+          echo "Reconfigure cabal projects to run tests for all dependencies"
+          sed -i 's|tests: False|tests: True|g' cabal.project
+        fi
```
Comment (angerman) [inline .github/workflows/haskell.yml:123]: I guess we want to keep the `cabal configure` line, and not do the cabal.project rewrite?

File: .github/workflows/haskell.yml [hunk 5]
```diff
@@ -1,73 +1,178 @@
...
+      uses: actions/cache@v2
       with:
-        path: ${{ steps.setup-haskell.outputs.cabal-store }}
-        key: cache-${{ runner.os }}-${{ matrix.ghc }}-v1-${{ hashFiles('cabal-cache.cabal') }}-${{ github.sha }}
-        restore-keys: cache-${{ runner.os }}-${{ matrix.ghc }}-v1-${{ hashFiles('cabal-cache.cabal') }}-
+        path: ${{ runner.os == 'Windows' && steps.win-setup-haskell.outputs.cabal-store || steps.lin-setup-haskell.outputs.cabal-store }}
+        key: cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}-${{ hashFiles('date.txt') }}
+        restore-keys: |
+          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
+          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}
```
Comment (angerman) [inline .github/workflows/haskell.yml:142]: These things really should have some `hash(env)` components. It's also missing in node and network :(

File: .github/workflows/haskell.yml (hunk 4 above)
Comment (erikd) [reply .github/workflows/haskell.yml:123]: That was replaced in a later version, where I took the CI definition from `node@master` and then fixed it for this project.

File: .github/workflows/haskell.yml (hunk 3 above)
Comment (erikd) [reply .github/workflows/haskell.yml:93]: That has been fixed.

**Comment by erikd** [comment]
Windows CI has been failing since April 2021. Do not consider that a blocker.

**Comment by angerman** [comment]
What? hydra... ohh my.

**Comment by angerman** [comment]
```
trace: WARNING: No sha256 found for source-repository-package https://github.com/input-output-hk/threepenny-gui ref=(unspecified) rev=c24c7effe8b2df011a31397bb409e560f3afb4ab download may fail in restricted mode (hydra)
```
ohh.

**Comment by michaelpj** [comment]
Since I assume we're going to push these to CHaP once this is merged, perhaps bump the package versions also while you're here?

**Comment by erikd** [comment]
Yes, was planning to do that one call `source-repository-package`s were removed.

**Comment by erikd** [comment]
Windows builds are still broken, and has been for 2 years https://github.com/input-output-hk/iohk-monitoring-framework/issues/604 . The problem seems to be in an upstream dependency.

**Pull request merged**